## Student grades prediction Project

**Prepared for FLIP ROBO by Olumide Ikumapayi-DS2306**

# About The Student grades Dataset

The dataset includes grades achieved by students during their time at university across different courses, as well as their CGPA, which is determined by their grades.For more information you can check the link to the dataset https://github.com/dsrscientist/dataset4/blob/main/Grades.csv

The flow of this machine learning exercise is as follows:

1. [Import Necessary Libraries](#Import-Necessary-Libraries)
2. [Importing the Student Grades Dataset](#Importing-the-Student-Grades-Dataset)
3. [Exploratory Data Analysis(EDA)](#Exploratory-Data-Analysis-(EDA))
4. [Data Preprocessing](#Data-Preprocessing)
5. [The Machine Learning Model Training,Evaluation,Predictions and Data Visualization](#The-Machine-Learning-Model-Training,-Evaluation,Predictions-and-Data-Visualization)
  * i.  [Linear Regression Model](#Linear-Regression-Model)
  * ii. [Random Forest Regressor](#Random-Forest-Regressor)
  * iii.[Support Vector Regression Model (SVR)](#Support-Vector-Regression-Model-(SVR))
  * iv. [Gradient Boosting Regressor](#Gradient-Boosting-Regressor)
  *  v. [Lasso Regression](#Lasso-Regression)
6. [Summary and Conclusion](#Summary-and-Conclusion)

# Import Necessary Libraries

Importing all necessary libraries for data manipulation, visualization, and machine learning(ML).

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.svm import SVR
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.linear_model import Lasso
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer

# Importing the Student Grades Dataset

In [2]:
url = "https://raw.githubusercontent.com/dsrscientist/dataset4/main/Grades.csv"
df = pd.read_csv("https://raw.githubusercontent.com/dsrscientist/dataset4/main/Grades.csv") 
df

Seat No. PH-121 HS-101 CY-105 HS-105/12 MT-111 CS-105 CS-106 EL-102  \
0    CS-97001     B-     D+     C-         C     C-     D+      D     C-   
1    CS-97002      A      D     D+         D     B-      C      D      A   
2    CS-97003      A      B      A        B-     B+      A     B-     B+   
3    CS-97004      D     C+     D+         D      D     A-     D+     C-   
4    CS-97005     A-     A-     A-        B+      A      A     A-     B+   
..        ...    ...    ...    ...       ...    ...    ...    ...    ...   
566  CS-97567      B      A      A        A-     A+      A     A-     A-   
567  CS-97568     A+      A      A         A      A      A      A     A-   
568  CS-97569      B      A     A-        B+      A      A      A      A   
569  CS-97570      A     B+      D         A      D     D+     B-     C-   
570  CS-97571      C      D      D         C      C     D+      B     C+   

    EE-119  ... CS-312 CS-317 CS-403 CS-421 CS-406 CS-414 CS-419 CS-423  \
0       B-  ...     C-     C-     C-     C-     A-      A     C-      B   
1       D+  ...     D+      D      C      D     A-     B-      C      C   
2       A-  ...      B      B      A      C      A      A      A     A-   
3        D  ...     D+      C     D+     C-     B-      B     C+     C+   
4        A  ...     B-     B+     B+     B-     A-      A     A-     A-   
..     ...  ...    ...    ...    ...    ...    ...    ...    ...    ...   
566     A+  ...     A-     A-      A      A      A     B+     B+      B   
567      A  ...     B+     B+      A      A     A-      B     A-      C   
568      A  ...     A-      B      A     B+      A      C     B+     A-   
569     B-  ...      D      B      B     C-      D      C      B     B-   
570      C  ...     C+      C     B-      D      F     C-     B+      D   

    CS-412   CGPA  
0       A-  2.205  
1        B  2.008  
2        A  3.608  
3       C+  1.906  
4        A  3.448  
..     ...    ...  
566      A  3.798  
567     A-  3.772  
568     A-  3.470  
569      C  2.193  
570     C-  1.753  

[571 rows x 43 columns]

# Exploratory Data Analysis(EDA)

Exploring the dataset in order to gain insights by Checking for missing values, summary statistics, and visualize the data.

In [3]:
df.head(15)

Seat No. PH-121 HS-101 CY-105 HS-105/12 MT-111 CS-105 CS-106 EL-102  \
0   CS-97001     B-     D+     C-         C     C-     D+      D     C-   
1   CS-97002      A      D     D+         D     B-      C      D      A   
2   CS-97003      A      B      A        B-     B+      A     B-     B+   
3   CS-97004      D     C+     D+         D      D     A-     D+     C-   
4   CS-97005     A-     A-     A-        B+      A      A     A-     B+   
5   CS-97006      A     B-      B        D+     C+     A-     C+      B   
6   CS-97007     B-     C-     A-         D     A-      B      A     B+   
7   CS-97008     B+     B+     C+         C      C     A-     C-     A-   
8   CS-97009     A-      C     A-        D+      B     A-     A-     A-   
9   CS-97010      A      B     C+         B      B     A-      C     B-   
10  CS-97011     A-     B-     B+        C-     C-     B-     C+     B-   
11  CS-97012     A-      C     A-        C+     C-      A     B-     B+   
12  CS-97013      B      C     A-         C     B-      B     D+      C   
13  CS-97014      A     C-     A-        C-      B     A-     C+     A-   
14  CS-97015      A      C     A-        C+     A-      A     B-     A-   

   EE-119  ... CS-312 CS-317 CS-403 CS-421 CS-406 CS-414 CS-419 CS-423 CS-412  \
0      B-  ...     C-     C-     C-     C-     A-      A     C-      B     A-   
1      D+  ...     D+      D      C      D     A-     B-      C      C      B   
2      A-  ...      B      B      A      C      A      A      A     A-      A   
3       D  ...     D+      C     D+     C-     B-      B     C+     C+     C+   
4       A  ...     B-     B+     B+     B-     A-      A     A-     A-      A   
5      B+  ...      C     B-     C+     C+     A-     A-     B+     B-      B   
6      A-  ...     B-      C     B+     B-      A     A-     B+     B+      A   
7       B  ...     D+     B+     B+      C      A     B+     A-     A-     A-   
8      B+  ...     B+      B     B-     C+      A     A-     B+     A-      A   
9      A-  ...     C+     B-     B+     B-      A      A     A-      B      A   
10      A  ...      C     B-     C-     C-      B     B+     B-     C-     B+   
11     A-  ...     A-     B-     B+      C      A     A-     B-     C+     A-   
12     B+  ...     C-      C     C-      C      A     A-     B-     A-     B+   
13     A-  ...      B     A-     A-     B+     A-      A      A     A-      A   
14     A-  ...     C+     A-     A-     B+     A-     A+      A      A      A   

     CGPA  
0   2.205  
1   2.008  
2   3.608  
3   1.906  
4   3.448  
5   3.026  
6   2.957  
7   3.043  
8   3.358  
9   3.247  
10  2.681  
11  3.281  
12  2.792  
13  3.523  
14  3.443  

[15 rows x 43 columns]

In [4]:
df.tail(30)

Seat No. PH-121 HS-101 CY-105 HS-105/12 MT-111 CS-105 CS-106 EL-102  \
541  CS-97542     A-     B+      A         A      A      A      A      B   
542  CS-97543      A     A-      A        A-      A      A     A-      B   
543  CS-97544      A      B      A        A-     A-      A     B+     B+   
544  CS-97545     A-     C-      A         A      B      B     A-      A   
545  CS-97546      B      B      A         A     A-     A-     A-      D   
546  CS-97547      B      D      A        B+     B+     A-     B-      D   
547  CS-97548     B-     C+     A-        B+     C-      B     C+      D   
548  CS-97549      A      A      A         A     C+      A     B+      C   
549  CS-97550     A-      C      A         A      B      B     B-     C-   
550  CS-97551     B-      B      A         C      A     A-     D+     A-   
551  CS-97552      A     B-      A        B+      A      A     B+     B+   
552  CS-97553     B-      C      A        B+     C-      A      B      C   
553  CS-97554      A      B     A+        C+      A      A     A-      A   
554  CS-97555     B+     C-     A+        C+      B     A-     B+      C   
555  CS-97556     B-      C      A        B-     C-      B      B     B-   
556  CS-97557      A     B+     A+         A      A     B+     A-      B   
557  CS-97558     B+     C+      B        A-     B-     B+     B+      C   
558  CS-97559     A-     B-      A         A     A-      A     B+     A-   
559  CS-97560     A-     C+      A         B      B     A-     C+      D   
560  CS-97561     B+     C-      A        D+     C+     A-     C-     C+   
561  CS-97562     B-      B     A-        B+     B-     B+     A-      A   
562  CS-97563      B      B      A         C      B     B+      B     C-   
563  CS-97564     A-      B      A        D+     B+     A-      C      A   
564  CS-97565     A-      D      A         C     B+     B+      C     A-   
565  CS-97566      F      F      D        C-    NaN     D+      D    NaN   
566  CS-97567      B      A      A        A-     A+      A     A-     A-   
567  CS-97568     A+      A      A         A      A      A      A     A-   
568  CS-97569      B      A     A-        B+      A      A      A      A   
569  CS-97570      A     B+      D         A      D     D+     B-     C-   
570  CS-97571      C      D      D         C      C     D+      B     C+   

    EE-119  ... CS-312 CS-317 CS-403 CS-421 CS-406 CS-414 CS-419 CS-423  \
541      A  ...      A      A     A-     A-     B+      B      A     B+   
542      A  ...     A-     A-     A+      A      A     B+      A     A-   
543     B+  ...     WU    NaN    NaN      W     WU    NaN    NaN    NaN   
544     A-  ...     B+      B     B+      B     B-     B+     A-     C+   
545     A-  ...      B      B      A     B+      B      C     B+     C-   
546      B  ...     B+     B+      A     B+     B+      B     B+     C+   
547     B-  ...      C     B-     A-     C+     D+     B-     B-      C   
548     B+  ...      B      B      B      B     B-     A-      A     C+   
549     B-  ...     B+     B-      A      B     B-      B      B     B-   
550      D  ...      A      C     B+     B-      C     D+      C      F   
551      A  ...     B+     B-      A      C     B+      C     B-      C   
552      B  ...     C+      C     B+      F     D+     C-      B      D   
553     B-  ...     A-     B+      A     B+      A     C+     A-      B   
554     B+  ...     A-     C+      A      B      B      B     B+     C-   
555     C-  ...      D     C-      B     C-     C+     B-     B+     D+   
556     A-  ...      A     A-      A     B+      A      B      A     B+   
557      B  ...     C-      C     A-      C     C+     D+      B     C-   
558     B+  ...     B+     B+      A      B      C      D     A-     D+   
559      B  ...     B+     B-      A     A-      A      C     B+     B+   
560     C+  ...     B+      C     B+     C-     C-      F      C      D   
561      B  ...     B+     B-     B+     D+     C+     B-     B+      C  

In [5]:
# Displaying the first few rows of the dataset
df.head()

Seat No. PH-121 HS-101 CY-105 HS-105/12 MT-111 CS-105 CS-106 EL-102 EE-119  \
0  CS-97001     B-     D+     C-         C     C-     D+      D     C-     B-   
1  CS-97002      A      D     D+         D     B-      C      D      A     D+   
2  CS-97003      A      B      A        B-     B+      A     B-     B+     A-   
3  CS-97004      D     C+     D+         D      D     A-     D+     C-      D   
4  CS-97005     A-     A-     A-        B+      A      A     A-     B+      A   

   ... CS-312 CS-317 CS-403 CS-421 CS-406 CS-414 CS-419 CS-423 CS-412   CGPA  
0  ...     C-     C-     C-     C-     A-      A     C-      B     A-  2.205  
1  ...     D+      D      C      D     A-     B-      C      C      B  2.008  
2  ...      B      B      A      C      A      A      A     A-      A  3.608  
3  ...     D+      C     D+     C-     B-      B     C+     C+     C+  1.906  
4  ...     B-     B+     B+     B-     A-      A     A-     A-      A  3.448  

[5 rows x 43 columns]

In [6]:
#Checking the dimension of the dataset
df.shape

(571, 43)

This dataset contains 571 rows and 43 columns.Out of which 1 is target variable and the remaining 42 are independent variables.

In [7]:
print(df.columns)

Index(['Seat No.', 'PH-121', 'HS-101', 'CY-105', 'HS-105/12', 'MT-111',
       'CS-105', 'CS-106', 'EL-102', 'EE-119', 'ME-107', 'CS-107', 'HS-205/20',
       'MT-222', 'EE-222', 'MT-224', 'CS-210', 'CS-211', 'CS-203', 'CS-214',
       'EE-217', 'CS-212', 'CS-215', 'MT-331', 'EF-303', 'HS-304', 'CS-301',
       'CS-302', 'TC-383', 'MT-442', 'EL-332', 'CS-318', 'CS-306', 'CS-312',
       'CS-317', 'CS-403', 'CS-421', 'CS-406', 'CS-414', 'CS-419', 'CS-423',
       'CS-412', 'CGPA'],
      dtype='object')


In [8]:
# Checking the columns of dataset
df.columns.tolist()

['Seat No.',
 'PH-121',
 'HS-101',
 'CY-105',
 'HS-105/12',
 'MT-111',
 'CS-105',
 'CS-106',
 'EL-102',
 'EE-119',
 'ME-107',
 'CS-107',
 'HS-205/20',
 'MT-222',
 'EE-222',
 'MT-224',
 'CS-210',
 'CS-211',
 'CS-203',
 'CS-214',
 'EE-217',
 'CS-212',
 'CS-215',
 'MT-331',
 'EF-303',
 'HS-304',
 'CS-301',
 'CS-302',
 'TC-383',
 'MT-442',
 'EL-332',
 'CS-318',
 'CS-306',
 'CS-312',
 'CS-317',
 'CS-403',
 'CS-421',
 'CS-406',
 'CS-414',
 'CS-419',
 'CS-423',
 'CS-412',
 'CGPA']

In [9]:
# checking the types of columns
df.dtypes

Seat No.      object
PH-121        object
HS-101        object
CY-105        object
HS-105/12     object
MT-111        object
CS-105        object
CS-106        object
EL-102        object
EE-119        object
ME-107        object
CS-107        object
HS-205/20     object
MT-222        object
EE-222        object
MT-224        object
CS-210        object
CS-211        object
CS-203        object
CS-214        object
EE-217        object
CS-212        object
CS-215        object
MT-331        object
EF-303        object
HS-304        object
CS-301        object
CS-302        object
TC-383        object
MT-442        object
EL-332        object
CS-318        object
CS-306        object
CS-312        object
CS-317        object
CS-403        object
CS-421        object
CS-406        object
CS-414        object
CS-419        object
CS-423        object
CS-412        object
CGPA         float64
dtype: object

There are two(2) types of data(float64 and object) present in the dataset.

In [10]:
#checking the null/missing values
df.isnull().sum()

Seat No.      0
PH-121        0
HS-101        0
CY-105        1
HS-105/12     1
MT-111        2
CS-105        0
CS-106        2
EL-102        2
EE-119        2
ME-107        2
CS-107        2
HS-205/20     5
MT-222        5
EE-222        7
MT-224        7
CS-210        7
CS-211        5
CS-203        5
CS-214        6
EE-217        6
CS-212        6
CS-215        6
MT-331        9
EF-303       10
HS-304       10
CS-301       10
CS-302       10
TC-383       10
MT-442       10
EL-332        9
CS-318        9
CS-306        9
CS-312       10
CS-317       12
CS-403       12
CS-421       12
CS-406       85
CS-414       13
CS-419       13
CS-423       14
CS-412       79
CGPA          0
dtype: int64

In [11]:
df.isnull().sum().sum()

425

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 571 entries, 0 to 570
Data columns (total 43 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Seat No.   571 non-null    object 
 1   PH-121     571 non-null    object 
 2   HS-101     571 non-null    object 
 3   CY-105     570 non-null    object 
 4   HS-105/12  570 non-null    object 
 5   MT-111     569 non-null    object 
 6   CS-105     571 non-null    object 
 7   CS-106     569 non-null    object 
 8   EL-102     569 non-null    object 
 9   EE-119     569 non-null    object 
 10  ME-107     569 non-null    object 
 11  CS-107     569 non-null    object 
 12  HS-205/20  566 non-null    object 
 13  MT-222     566 non-null    object 
 14  EE-222     564 non-null    object 
 15  MT-224     564 non-null    object 
 16  CS-210     564 non-null    object 
 17  CS-211     566 non-null    object 
 18  CS-203     566 non-null    object 
 19  CS-214     565 non-null    object 
 20  EE-217    

In [13]:
# Checking if 'Seat No' column exists before dropping it
if 'Seat No.' in df.columns:
    # Drop the 'Seat No' column as it's not needed for modeling
    df.drop('Seat No.', axis=1, inplace=True)


In [14]:
# One-hot encode the course grade columns
df = pd.get_dummies(df, columns=df.columns[2:], drop_first=True)

In [15]:
# droping rows with missing values
df.dropna(inplace=True)

In [16]:
# Verifying that there are no missing values left
missing_values_after_handling = df.isnull().sum()

In [17]:
# Summary statistics
df.describe()

CY-105_A+   CY-105_A-    CY-105_B   CY-105_B+   CY-105_B-    CY-105_C  \
count  571.000000  571.000000  571.000000  571.000000  571.000000  571.000000   
mean     0.054291    0.210158    0.085814    0.087566    0.073555    0.033275   
std      0.226789    0.407778    0.280335    0.282910    0.261274    0.179511   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

        CY-105_C+   CY-105_C-    CY-105_D   CY-105_D+  ...  CGPA_3.919  \
count  571.000000  571.000000  571.000000  571.000000  ...  571.000000   
mean     0.029772    0.028021    0.054291    0.024518  ...    0.003503   
std      0.170108    0.165178    0.226789    0.154788  ...    0.059131   
min      0.000000    0.000000    0.000000    0.000000  ...    0.000000   
25%      0.000000    0.000000    0.000000    0.000000  ...    0.000000   
50%      0.000000    0.000000    0.000000    0.000000  ...    0.000000   
75%      0.000000    0.000000    0.000000    0.000000  ...    0.000000   
max      1.000000    1.000000    1.000000    1.000000  ...    1.000000   

       CGPA_3.925  CGPA_3.942  CGPA_3.943  CGPA_3.946  CGPA_3.949  CGPA_3.963  \
count  571.000000  571.000000  571.000000  571.000000  571.000000  571.000000   
mean     0.001751    0.003503    0.001751    0.001751    0.001751    0.001751   
std      0.041849    0.059131    0.041849    0.041849    0.041849    0.041849   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

        CGPA_3.97  CGPA_3.972  CGPA_3.985  
count  571.000000  571.000000  571.000000  
mean     0.001751    0.001751    0.001751  
std      0.041849    0.041849    0.041849  
min      0.000000    0.000000    0.000000  
25%      0.000000    0.000000    0.000000  
50%      0.000000    0.000000    0.000000  
75%      0.000000    0.000000    0.000000  
max      1.000000    1.000000    1.000000  

[8 rows x 952 columns]

**This gives statistical information of the numerical columns.The summary of the dataset looks perfect since there is no negative/invalid values present.**

The counts of all the columns are the same which means there are no missing values in the dataset.

In [18]:
print(df.isnull().sum())

PH-121        0
HS-101        0
CY-105_A+     0
CY-105_A-     0
CY-105_B      0
             ..
CGPA_3.949    0
CGPA_3.963    0
CGPA_3.97     0
CGPA_3.972    0
CGPA_3.985    0
Length: 954, dtype: int64


In [19]:
df.describe().T

count      mean       std  min  25%  50%  75%  max
CY-105_A+   571.0  0.054291  0.226789  0.0  0.0  0.0  0.0  1.0
CY-105_A-   571.0  0.210158  0.407778  0.0  0.0  0.0  0.0  1.0
CY-105_B    571.0  0.085814  0.280335  0.0  0.0  0.0  0.0  1.0
CY-105_B+   571.0  0.087566  0.282910  0.0  0.0  0.0  0.0  1.0
CY-105_B-   571.0  0.073555  0.261274  0.0  0.0  0.0  0.0  1.0
...           ...       ...       ...  ...  ...  ...  ...  ...
CGPA_3.949  571.0  0.001751  0.041849  0.0  0.0  0.0  0.0  1.0
CGPA_3.963  571.0  0.001751  0.041849  0.0  0.0  0.0  0.0  1.0
CGPA_3.97   571.0  0.001751  0.041849  0.0  0.0  0.0  0.0  1.0
CGPA_3.972  571.0  0.001751  0.041849  0.0  0.0  0.0  0.0  1.0
CGPA_3.985  571.0  0.001751  0.041849  0.0  0.0  0.0  0.0  1.0

[952 rows x 8 columns]

In [20]:
# Defining the grade mapping dictionary
grade_mapping = {
    'A': 4.0,
    'B': 3.0,
    'C': 2.0,
    'D': 1.0,
    'F': 0.0
}

In [21]:
# Converting grades to numerical values in the dataset
for column in df.columns:
    if column != 'CGPA':
        df[column] = df[column].map(grade_mapping)

# Data Preprocessing

In [22]:
# Defining the target variable
target_column = 'CGPA'

In [23]:
# Printing the column names of the dataset
print(df.columns)

Index(['PH-121', 'HS-101', 'CY-105_A+', 'CY-105_A-', 'CY-105_B', 'CY-105_B+',
       'CY-105_B-', 'CY-105_C', 'CY-105_C+', 'CY-105_C-',
       ...
       'CGPA_3.919', 'CGPA_3.925', 'CGPA_3.942', 'CGPA_3.943', 'CGPA_3.946',
       'CGPA_3.949', 'CGPA_3.963', 'CGPA_3.97', 'CGPA_3.972', 'CGPA_3.985'],
      dtype='object', length=954)


In [24]:
# Filling missing values with the mean of the respective columns
df = df.fillna(df.mean())

In [25]:
# Checking for non-numeric values in the dataset
non_numeric_values = df.select_dtypes(exclude=[np.number])
print("Non-Numeric Values:")
print(non_numeric_values)

Non-Numeric Values:
Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, ...]

[571 rows x 0 columns]


In [26]:
# Printing the first few rows of the DataFrame
print(df.head())

     PH-121    HS-101  CY-105_A+  CY-105_A-  CY-105_B  CY-105_B+  CY-105_B-  \
0  2.947154  2.427907        NaN        NaN       NaN        NaN        NaN   
1  4.000000  1.000000        NaN        NaN       NaN        NaN        NaN   
2  4.000000  3.000000        NaN        NaN       NaN        NaN        NaN   
3  1.000000  2.427907        NaN        NaN       NaN        NaN        NaN   
4  2.947154  2.427907        NaN        NaN       NaN        NaN        NaN   

   CY-105_C  CY-105_C+  CY-105_C-  ...  CGPA_3.919  CGPA_3.925  CGPA_3.942  \
0       NaN        NaN        NaN  ...         NaN         NaN         NaN   
1       NaN        NaN        NaN  ...         NaN         NaN         NaN   
2       NaN        NaN        NaN  ...         NaN         NaN         NaN   
3       NaN        NaN        NaN  ...         NaN         NaN         NaN   
4       NaN        NaN        NaN  ...         NaN         NaN         NaN   

   CGPA_3.943  CGPA_3.946  CGPA_3.949  CGPA_3.963  CGPA_

# Spliting the data into training and testing sets.

In [93]:
# Defining the target variable (CGPA) and feature columns (course grades)
target_column = 'CGPA_3.985'  

In [94]:
# Extracting the target variable (CGPA)
y = df[target_column]

In [27]:
# Extracting the feature columns (course grades)
feature_columns = df.columns[df.columns.str.match(r'[A-Z]{2}-\d{3}_\d\.\d')]

In [33]:
# Initializing the imputer with a strategy to fill missing values with the mean
imputer = SimpleImputer(strategy='mean')